metrics/README.md

```diff
@@ -0,0 +1,90 @@
+downstream dependents, and we'd like all the code shipped in `knative.dev/pkg`
+to be able to be tested without needing any special environment setup.
+
+With the current direct-integration setup, there needs to be initial and ongoing
+work in `pkg` (which should be high-value, low-churn code) to maintain and
+update stats exporters which need to be statically linked into ~all Knative
+binaries. This setup also causes problems for vendors who may want or need to
+perform an out-of-tree integration (e.g. proprietary or partially-proprietary
+monitoring stacks).
+
+Another problem is that each vendor's exporter requires different parameters,
+supplied as Golang `Options` methods which may require complex connections with
+the Knative ConfigMap. Two examples of this are secrets like API keys and the
+Prometheus monitoring port (which requires additional service/etc wiring).
+
+See also
+[this doc](https://docs.google.com/document/d/1t-aov3XrhobjCKW4kwScY44QAoahiwxoyXXFtZyL8jw/edit),
+where the plan was worked out.
+
+## The plan
+
+OpenCensus (and eventually OpenTelemetry) offers an sidecar or host-level agent
+with speaks the OpenCensus protocol and can proxy from this protocol to multiple
+backends.
+
+![OpenCensus Agent configuration](https://github.com/census-instrumentation/opencensus-service/raw/master/images/opencensus-service-deployment-models.png)
+
+**We will standardize on export to the OpenCensus agent, and encourage vendors
+to implement their own OpenCensus DaemonSet which connects to their desired
+monitoring environment.**
+
+**Additionally, once OpenTelemetry agent is stable, we will propose adding the
+OpenTelemetry agent running on a localhost port as part of the runtime
+contract.**
+
+We need to make sure that the OpenCensus library does not block, fail, or queue
+metrics in-process excessively in the case where the OpenCensus Agent is not
+present on the cluster. This will allow us to ship Knative components which
+attempt to reach out the Agent if present, and which simply retain local
+statistics for a short period of time if not.
+
+### Concerns
+
+- Unsure about the stability of the OpenCensus Agent (or successor). We're
+  currently investingating this, but the OpenCensus agent seems to have been
+  recommended by several others.
+- Running `fluentd` as a sidecar was very big (400MB) and had a large impact on
+  cold start times.
+  - Mitigation: run the OpenCensus agent as a DaemonSet (like we do with
+    `fluentd` now).
+- Running as a DaemonSet may make it more difficult to ensure that metrics for
+  each namespace end up in the right place.
+  - We have this problem with the built-in configurations today, so this doesn't
+    make the problem substantially worse.
+  - May want/need some connection between the Agent and the Kubelet to verify
+    sender identities eventually.
+  - Only expose OpenCensus Agent on localhost, not outside the node.
+
+### Steps to reach the goal
```

Comment:
We discussed putting a deadline on this to prevent an infinite creep, which will influence how the AI issues tile into milestones.  Do we want to incorporate that here at all?

Reply:
This also doesn't surface our level of tolerance for featurework in the short-term, now that we have this plan.

Reply:
I put a cutoff in at 0.13, since the next two weeks are probably out for most people.